File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/InventoryQuery/InventoryQueryDecoratorBase.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.InventoryQuery;

/// <summary>
///     Base class for decorators that add filtering or sorting behavior to an <see cref="IInventoryQuery" />.
/// </summary>
public abstract class InventoryQueryDecoratorBase : IInventoryQuery
{
    /// <summary>
    ///     Initializes a new instance of the <see cref="InventoryQueryDecoratorBase" /> class.
    /// </summary>
    /// <param name="inner">The inner query to decorate.</param>
    protected InventoryQueryDecoratorBase(IInventoryQuery inner)
    {
        Inner = inner ?? throw new ArgumentNullException(nameof(inner));
    }

    /// <summary>
    ///     Gets the inner query being decorated.
    /// </summary>
    private IInventoryQuery Inner { get; }

    /// <summary>
    ///     Gets the description of the filter or sort applied by this decorator.
    /// </summary>
    protected virtual string? AppliedDescription => null;

    /// <inheritdoc />
    public IReadOnlyList<InventoryItem> Execute()
    {
        var items = Inner.Execute();
        return Decorate(items);
    }

    /// <inheritdoc />
    public IReadOnlyList<string> AppliedFiltersAndSorts
    {
        get
        {
            if (AppliedDescription is null)
                return Inner.AppliedFiltersAndSorts;

            var combined = new List<string>(Inner.AppliedFiltersAndSorts.Count + 1);
            combined.AddRange(Inner.AppliedFiltersAndSorts);
            combined.Add(AppliedDescription);
            return combined.AsReadOnly();
        }
    }

    /// <summary>
    ///     Applies the decorator's specific logic (filtering or sorting) to the results of the inner query.
    /// </summary>
    /// <param name="items">The items returned by the inner query.</param>
    /// <returns>the decorated (filtered or sorted) list of items.</returns>
    protected abstract IReadOnlyList<InventoryItem> Decorate(IReadOnlyList<InventoryItem> items);
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventoryquery/InventoryQueryDecoratorBase.java
package assignment2solution.domain.inventoryquery;

import assignment2solution.domain.inventory.InventoryItem;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Base class for decorators that add filtering or sorting behavior to an {@link IInventoryQuery}.
 */
public abstract class InventoryQueryDecoratorBase implements IInventoryQuery {
    private final IInventoryQuery inner;

    /**
     * Initializes a new instance of the {@link InventoryQueryDecoratorBase} class.
     *
     * @param inner The inner query to decorate.
     */
    protected InventoryQueryDecoratorBase(IInventoryQuery inner) {
        this.inner = Objects.requireNonNull(inner, "inner");
    }

    /**
     * Gets the description of the filter or sort applied by this decorator.
     */
    protected String getAppliedDescription() {
        return null;
    }

    @Override
    public List<InventoryItem> execute() {
        var items = inner.execute();
        return decorate(items);
    }

    @Override
    public List<String> getAppliedFiltersAndSorts() {
        var description = getAppliedDescription();
        if (description == null) {
            return inner.getAppliedFiltersAndSorts();
        }

        var combined = new ArrayList<String>(inner.getAppliedFiltersAndSorts().size() + 1);
        combined.addAll(inner.getAppliedFiltersAndSorts());
        combined.add(description);
        return List.copyOf(combined);
    }

    /**
     * Applies the decorator's specific logic (filtering or sorting) to the results of the inner query.
     *
     * @param items The items returned by the inner query.
     * @return the decorated (filtered or sorted) list of items.
     */
    protected abstract List<InventoryItem> decorate(List<InventoryItem> items);
}
